### PR TITLE
ROX-18568: Prune downloadable report

### PR DESCRIPTION
**Pull request by c-du**
## Description

A detailed explanation of the changes in your PR.

Feel free to remove this section if it is overkill for your PR, and the title of your PR is sufficiently descriptive.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] ~Evaluated and added CHANGELOG entry if required~
- [ ] ~Determined and documented upgrade steps~
- [ ] ~Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed
Add unit test
TODO(replace-me): Manual e2e test.

In addition to reviewing your code, reviewers **must** also review your testing
instructions and make sure they are sufficient.

### CODE REVIEW

File: proto/storage/config.proto [hunk 1]
```diff
@@ -55,8 +55,11 @@ message DecommissionedClusterRetentionConfig {
     google.protobuf.Timestamp created_at       = 4;
 }
 
+// next available tag: 4
 message ReportRetentionConfig {
-    int32 history_retention_duration_days       = 1;
+    int32 history_retention_duration_days              = 1;
+    uint32 downloadable_report_retention_days          = 2;
```
Comment (c-du) [inline proto/storage/config.proto:61]: It looks like uint32 should be better. We could make a dedicated PR to change if needed.

Comment (md2119) [reply proto/storage/config.proto:61]: That makes sense since the number of days is >= 0

File: central/pruning/pruning.go [hunk 2]
```diff
@@ -567,6 +574,46 @@ func (g *garbageCollectorImpl) removeOldReportHistory(config *storage.PrivateCon
...
+	query := search.NewQueryBuilder().AddRegexes(search.BlobName, common.ReportBlobRegex).ProtoQuery()
+	blobs, err := g.blobStore.SearchMetadata(pruningCtx, query)
+	if err != nil {
+		log.Errorf("Failed to fetch downloadable report metadata: %v", err)
+		return
+	}
+	// Sort reversely by modification time
+	sort.Slice(blobs, func(i, j int) bool {
+		return blobs[i].GetModifiedTime().Compare(blobs[j].GetModifiedTime()) > 0
+	})
```
Comment (md2119) [inline central/pruning/pruning.go:594]: Can you add sort option to the query? The `modified_time` field is sortable using `Blob Modified On` search label.

File: central/pruning/pruning.go [hunk 3]
```diff
@@ -567,6 +573,45 @@ func (g *garbageCollectorImpl) removeOldReportHistory(config *storage.PrivateCon
...
+	blobRetentionDays := config.GetReportRetentionConfig().GetDownloadableReportRetentionDays()
+	cutOffTime, err := types.TimestampProto(time.Now().Add(-time.Duration(blobRetentionDays) * 24 * time.Hour))
+	if err != nil {
+		log.Errorf("Failed to determine downloadable report retention %v", err)
+		return
+	}
+
+	// Sort reversely by modification time
+	query := search.NewQueryBuilder().AddRegexes(search.BlobName, common.ReportBlobRegex).WithPagination(
+		search.NewPagination().AddSortOption(search.NewSortOption(search.BlobModificationTime).Reversed(true)))
```
Comment (md2119) [inline central/pruning/pruning.go:586]: Why are you order the blobs last modified (latest) first?

**Review by md2119 — APPROVED** [review]
Just one clarifying question.

File: central/pruning/pruning.go (hunk 3 above)
Comment (c-du) [reply central/pruning/pruning.go:586]: From the latest, I took blobs and counting them until size quota filled or cut off time reached, then kill the rest.
